openedx/features/calendar_sync/ics.py

```diff
@@ -0,0 +1,73 @@
+""" Generate .ics files from a user schedule """
+
+from datetime import datetime, timedelta
+
+import pytz
+from django.conf import settings
+from icalendar import Calendar, Event, vCalAddress, vText
+
+from lms.djangoapps.courseware.courses import get_course_assignments
+from openedx.core.djangoapps.site_configuration.helpers import get_value
+from openedx.core.djangolib.markup import HTML
+
+from . import get_calendar_event_id
+
+
+def generate_ics_for_event(uid, summary, url, now, start, organizer_name, organizer_email):
+    """
+    Generates an ics-formatted bytestring for the given assignment information.
+
+    To pretty-print the bytestring, do: `ics.decode('utf8').replace('\r\n', '\n')`
+    """
+    # icalendar library: https://icalendar.readthedocs.io/en/latest/
+    # ics format spec: https://tools.ietf.org/html/rfc2445
+    # ics conventions spec: https://tools.ietf.org/html/rfc5546
+
+    organizer = vCalAddress('mailto:' + organizer_email)
+    organizer.params['cn'] = vText(organizer_name)
+
+    event = Event()
+    event.add('uid', uid)
+    event.add('dtstamp', now)
+    event.add('organizer', organizer, encode=0)
+    event.add('summary', summary)
+    # FIXME description should be translated if we use hardcoded text, once we finalize that text
+    event.add('description', HTML('<a href="{url}">Link</a>').format(url=url))
+    event.add('dtstart', start)
+    event.add('duration', timedelta(0))
+    event.add('transp', 'TRANSPARENT')  # available, rather than busy
+
+    cal = Calendar()
+    cal.add('prodid', '-//Open edX//calendar_sync//EN')
```

Comment:
Should this vary per open-edx provider?

Reply:
I figured no. I used the string "Open edX" as the product name to specifically avoid having to do that.

The event ids do use `@hostname` to globally differentiate themselves from other open edx providers. But I figured the product id, which is never visible to users, doesn't need to be unique per provider.

Reply:
Ok, great. I just wasn't sure if we needed anything to disambiguate platforms.

Reply:
Naw, seems to be more of a "what software made this" kind of thing. Like user agent in http.